plutus-tx/src/PlutusTx/Data/AssocMap.hs

```diff
@@ -0,0 +1,413 @@
+                v' = case lookup' k ls of
+                  Just r ->
+                    P.toBuiltinData
+                      ( These
+                          (P.unsafeFromBuiltinData v)
+                          (P.unsafeFromBuiltinData r)
+                        :: These a b
+                      )
+                  Nothing ->
+                    P.toBuiltinData (That (P.unsafeFromBuiltinData v) :: These a b)
+             in BI.mkCons (BI.mkPairData k v') (goRight tl)
+        )
+
+    res = goLeft ls `safeAppend` goRight rs
+
+    safeAppend xs1 xs2 =
+      P.matchList
+        xs1
+        (\() -> xs2)
+        ( \hd tl ->
+            let k = BI.fst hd
+                v = BI.snd hd
+             in insert' k v (safeAppend tl xs2)
+        )
+
+-- | Combine two 'Map's with the given combination function.
+unionWith ::
+  forall k a.
+  (P.UnsafeFromData a, P.ToData a) =>
+  (a -> a -> a) ->
+  Map k a ->
+  Map k a ->
+  Map k a
+unionWith f (Map ls) (Map rs) =
+  Map res
+  where
+    ls' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    ls' = go ls
+      where
+        go xs =
+          P.matchList
+            xs
+            (\() -> nil)
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    v' = BI.snd hd
+                    v'' = case lookup' k' rs of
+                      Just r ->
+                        P.toBuiltinData
+                          (f (P.unsafeFromBuiltinData v') (P.unsafeFromBuiltinData r))
+                      Nothing -> v'
+                 in BI.mkCons (BI.mkPairData k' v'') (go tl)
+            )
+
+    rs' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    rs' = go rs
+      where
+        go xs =
+          P.matchList
+            xs
+            (\() -> nil)
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    tl' = go tl
+                 in if member' k' ls
+                      then tl'
+                      else BI.mkCons hd tl'
+            )
+
+    res :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    res = go rs' ls'
+      where
+        go acc xs =
+          P.matchList
+            xs
+            (\() -> acc)
+            (\hd -> go (BI.mkCons hd acc))
+
+{-# INLINEABLE toList #-}
+-- | Convert the `Map` to a list of key-value pairs. This operation is O(n).
+-- See 'toBuiltinList' for a more efficient alternative.
+toList :: (P.UnsafeFromData k, P.UnsafeFromData a) => Map k a -> [(k, a)]
+toList d = go (toBuiltinList d)
+  where
+    go xs =
+      P.matchList
+        xs
+        (\() -> [])
+        ( \hd tl ->
+            (P.unsafeFromBuiltinData (BI.fst hd), P.unsafeFromBuiltinData (BI.snd hd))
+              : go tl
+        )
+
+{-# INLINEABLE toBuiltinList #-}
+-- | Convert the `Map` to a `P.BuiltinList` of key-value pairs. This operation is O(1).
+toBuiltinList :: Map k a -> BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+toBuiltinList (Map d) = d
```

Comment:
For the tests that we currently have it doesn't make any difference. I was thinking it'd be nice to have it as a separate function because it's easier document this way but maybe that doesn't make sense. In any case I can tweak this in one of the follow-up PRs.